crypto/cryptlib.c

```diff
@@ -355,4 +355,54 @@ int CRYPTO_memcmp(const void * in_a, const void * in_b, size_t len)
 
     return x;
 }
+
+/*
+ * For systems that don't provide an instruction counter register or equivalent
+ * we provide a best available clock.
+ * Return zero if no such timer is available.
+ */
+uint32_t OPENSSL_rdtsc(void)
```

Comment:
Since we have assembler versions of this function, this will most likely cause at least a duplicate symbol warning, this should probably be guarded by OPENSSL_CPUID_OBJ

But I think even that is not enough. There are platforms where we have assembler versions of that function, but where the processor either doesn't support or, or does not allow it from user space.

The easiest workaround is that either you rename this function, or all the assembler functions, and have this function call the assembler version and check that the return value != 0. And if it's 0, do one of the things below.

Reply:
It *is* guarded, see line 336.

Reply:
Also, I'm not exactly sure what you're looking at.  The mechanism is the `cpuid_asm_src` attribute in the diverse config targets, they are usually inherited from a set of templates in `Configurations/00-base-templates.conf`, and `OPENSSL_CPUID_OBJ` is only defined if the `cpuid_asm_src` attribute isn't `mem_clr.c` (or the corresponding `cpuid_obj` is `mem_clr.o`).  So there's a direct relationship between the definition of `OPENSSL_CPUID_OBJ` and the assembler implementations.  A symbol clash is therefore very unlikely.

Reply:
I don't understand either.  I thought I had guarded it properly.  I used the same guard as calls to `OPENSSL_rdtsc` used.

Reply:
I missed it because it wasn't shown in the diff.

Anyway, since the intention is to use this as a source for the additional data in the RNG, I would really like that it actually tries to get a value which is not always the case with the assembler routines.